File: lua/KDampedOscillator.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
if (not KDampedOscillator) then
KDampedOscillator	=	{ }

KDampedOscillator.className			 	= 	'kDampedOscillator'

KDampedOscillator.k_strongly_damped 	= 	0
KDampedOscillator.k_weakly_damped		=	1
KDampedOscillator.k_critically_damped	=	2

--						Based on my old MST204 math course notes.
--						m_ 				=	0,
--						r_ 				=	0,
--						k_ 				=	0,
--	
--						x0_				=	0,
--						v0_				=	0,
--						xe_ 			=	0,
--	
						-- Natural angular frequency and damping ratio.
--						v_ 
--						w_ 
--						alpha_ 
--
						-- Coefficients speficic to strong damping.
--						private float	b_, c_ 
--						private float	l1_, l2_
						-- Coefficients specific to weak damping.
--						private float	a_ 
--						private float 	p_, phi_ 

--
--	kdampedoscillator()	-	default constructor
--
--	Initialise the spring constants to something...
--
--	public KDampedOscillator()
--	{
--		this(1.0f,1.0f,100.0f,0.0f,0.0f) 
--	} 
	
--	public KDampedOscillator(float m, float r, float k)
--	{
--		this(m,r,k,0.0f,0.0f) 
--	}

--m=1 r=0 k=100 will make it shake very fast

	function KDampedOscillator.Create( m,  r,  k,  x,  v)
		local kdampedoscillator = {
			x0_ 		= x, 
			v0_			= v,
			m_			= m,
			r_			= r,
			k_  		= k,
			xe_			= 0,
			v_			= 0,
			w_			= 0,
		    alpha 		= 0,
			p_			= 0,
			className	= KDampedOscillator.className,
		}
		setmetatable(kdampedoscillator,{__index = KDampedOscillator})	
		return kdampedoscillator
	end
	
	
	-- Reference MST204/MST209 Handbook
	function KDampedOscillator.reset(kdampedoscillator, x0,  v0,  xe)
		
		kdampedoscillator.x0_ = x0 - xe 
		kdampedoscillator.v0_ = v0 
		kdampedoscillator.xe_ = xe 

		-- Calculate the natural angular frequency and damping ratio.

		kdampedoscillator.w_ = math.sqrt( kdampedoscillator.k_ / kdampedoscillator.m_ ) 
		kdampedoscillator.alpha_ = kdampedoscillator.r_ / ( 2 * math.sqrt( kdampedoscillator.m_ * kdampedoscillator.k_ ) ) 

	 	-- Calculate the solution type.

		if ( kdampedoscillator.alpha_ > 1 ) then
			kdampedoscillator.damping_type_ = KDampedOscillator.k_strongly_damped 
		elseif ( kdampedoscillator.alpha_ < 1 ) then
			kdampedoscillator.damping_type_ = KDampedOscillator.k_weakly_damped 
		else
			kdampedoscillator.damping_type_ = KDampedOscillator.k_critically_damped 
		end
		--assert(false,'DT = '..kdampedoscillator.damping_type_)	
	 	-- Solve the equation of motion.
		
		local damping_type_ = kdampedoscillator.damping_type_ 
		
			
		if (damping_type_ ==  KDampedOscillator.k_strongly_damped) then
		
			local a = math.sqrt( kdampedoscillator.alpha_*kdampedoscillator.alpha_ - 1.0) 

			kdampedoscillator.l1_ = kdampedoscillator.w_*( -kdampedoscillator.alpha_ + a ) 
			kdampedoscillator.l2_ = kdampedoscillator.w_*( -kdampedoscillator.alpha_ - a ) 

			-- Initial conditions.

			kdampedoscillator.c_ = ( kdampedoscillator.v0_ - kdampedoscillator.l1_*kdampedoscillator.x0_ ) / ( kdampedoscillator.l2_ - kdampedoscillator.l1_ ) 
			kdampedoscillator.b_ = kdampedoscillator.x0_ - kdampedoscillator.c_ 
			
		elseif (damping_type_ ==  KDampedOscillator.k_weakly_damped) then
			
			-- x(t) = A*exp(-pt)*cos(vt+phi)

			kdampedoscillator.p_ = kdampedoscillator.r_ / kdampedoscillator.m_ * 0.5 
			kdampedoscillator.v_ = math.sqrt( 4.0*kdampedoscillator.m_*kdampedoscillator.k_ - kdampedoscillator.r_*kdampedoscillator.r_ ) / kdampedoscillator.m_ * 0.5
			kdampedoscillator.a_ = math.sqrt( kdampedoscillator.x0_*kdampedoscillator.x0_ + kdampedoscillator.v0_*kdampedoscillator.v0_ / (kdampedoscillator.p_*kdampedoscillator.p_))

			if (kdampedoscillator.a_ == 0.0) then
				kdampedoscillator.phi_ =	0.0
			else
				kdampedoscillator.phi_ =	math.acos( kdampedoscillator.x0_ / kdampedoscillator.a_ ) 
			end
		
		elseif (damping_type_ ==  KDampedOscillator.k_critically_damped) then

			kdampedoscillator.c_ = kdampedoscillator.x0_ 
			kdampedoscillator.b_ = kdampedoscillator.v0_ + kdampedoscillator.x0_ * kdampedoscillator.w_ 
		end
			
			
		
		
	end
	
	
	-- 
	-- 	centre()	-	returns the centre position of the spring
	-- 
	-- 	The centre is the position when the spring is in equilibrium.
	-- 
	-- public void centre( float & x ) -- const
	-- {
	-- 	x = xe_ 
	-- }
	
	function KDampedOscillator.centre(kdampedoscillator) 
		return kdampedoscillator.xe_ 
	end
	

	-- 
	-- 	equilibrium()	-	check if the oscillator is in static equilibrium
	-- 
	-- 	The oscillator is static if it is at the equilibrium position with zero velocity.
	-- 
	-- 	Parameters:
	-- 
	-- 		t		-	the time to evaluate the system at
	-- 		epsilon	-	how near to true equilbrium to calculate to
	-- 
	-- 	Returns:
	-- 
	-- 		the position of the system at the time t
	-- 

	function KDampedOscillator.equilibrium(kdampedoscillator, t,  epsilon ) 
		return math.abs( KDampedOscillator.evaluate(kdampedoscillator, t ) - kdampedoscillator.xe_ ) <= epsilon and math.abs( KDampedOscillator.velocity(kdampedoscillator, t ) ) <= epsilon 
	end
	
	
	
	function KDampedOscillator.evaluate(kdampedoscillator, t)

		local x = 0.0
		local damping_type_ = kdampedoscillator.damping_type_ 
		
		if (damping_type_ == KDampedOscillator.k_strongly_damped) then
			x = kdampedoscillator.b_*math.exp( kdampedoscillator.l1_*t ) + kdampedoscillator.c_*math.exp( kdampedoscillator.l2_*t ) 
		elseif (damping_type_ == KDampedOscillator.k_weakly_damped) then
			x = kdampedoscillator.a_*math.exp( -kdampedoscillator.p_*t ) * math.cos( kdampedoscillator.v_*t + kdampedoscillator.phi_ ) 
		elseif (damping_type_ == KDampedOscillator.k_critically_damped) then
			x = (kdampedoscillator.b_*t +kdampedoscillator.c_)*math.exp( -kdampedoscillator.w_*t ) 
		end			
		
		return kdampedoscillator.xe_ + x
		
	end
	
	function KDampedOscillator.velocity(kdampedoscillator, t)

		local v = 0.0
		local damping_type_ = kdampedoscillator.damping_type_ 
		if (damping_type_ == KDampedOscillator.k_strongly_damped) then
				v = kdampedoscillator.b_*kdampedoscillator.l1_*math.exp( kdampedoscillator.l1_*t ) + kdampedoscillator.c_*kdampedoscillator.l2_*math.exp( kdampedoscillator.l2_*t ) 
		elseif (damping_type_ == KDampedOscillator.k_weakly_damped) then
				v = -kdampedoscillator.a_*kdampedoscillator.p_*math.exp( -kdampedoscillator.p_*t ) * math.cos( kdampedoscillator.v_*t + kdampedoscillator.phi_ ) - kdampedoscillator.v_ * kdampedoscillator.a_*math.exp( -kdampedoscillator.p_*t ) * math.sin( kdampedoscillator.v_*t + kdampedoscillator.phi_ ) 
		elseif (damping_type_  == KDampedOscillator.k_critically_damped) then
				v = ( kdampedoscillator.b_ - kdampedoscillator.w_*(kdampedoscillator.b_*t +kdampedoscillator.c_ ) )*math.exp( -kdampedoscillator.w_*t ) 
		end		

		return v 
	end
	
end

File: source/library/KDampedOscillator.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

// Based on Damped Oscillator Maths Unit MST204/MST209 Open University

var k_strongly_damped 	= 	0
var k_weakly_damped		=	1
var k_critically_damped	=	2

function KDampedOscillator( m,  r,  k,  x,  v)
{
			this.x0_ 		= x ; 
			this.v0_		= v ;
			this.m_			= m ;
			this.r_			= r ;
			this.k_  		= k ;
			this.xe_		= 0 ;
			this.v_			= 0 ;
			this.w_			= 0 ;
		    this.alpha 		= 0 ;
			this.p_			= 0 ;
}


KDampedOscillator.Create				=	function( m,  r,  k,  x,  v)
											{
												return new KDampedOscillator( m,  r,  k,  x,  v) ;
											}

KDampedOscillator.prototype = 
{
	constructor: KDampedOscillator,


	reset: function( x0,  v0,  xe) 
	{
		
		this.x0_ = x0 - xe ;
		this.v0_ = v0 ;
		this.xe_ = xe ;

		// Calculate the natural angular frequency and damping ratio.

		this.w_ 	= Math.sqrt( this.k_ / this.m_ ) ;
		this.alpha_ = this.r_ / ( 2 * Math.sqrt( this.m_ * this.k_ ) ) ;

	 	// Calculate the solution type.

		if ( this.alpha_ > 1 ) {
			this.damping_type_ = k_strongly_damped ;
		} else if ( this.alpha_ < 1 ) {
			this.damping_type_ = k_weakly_damped ;
		} else {
			
			this.damping_type_ = k_critically_damped ;
		}
		var damping_type_ = this.damping_type_ ;

		switch(damping_type_)
		{
			case k_strongly_damped:
				var a = Math.sqrt( this.alpha_*this.alpha_ - 1.0) ;

				this.l1_ = this.w_*( -this.alpha_ + a ) ;
				this.l2_ = this.w_*( -this.alpha_ - a ) ;

			// Initial conditions.

				this.c_ = ( this.v0_ - this.l1_*this.x0_ ) / ( this.l2_ - this.l1_ ) ;
				this.b_ = this.x0_ - this.c_ ;
				
				break ;
				
			case k_weakly_damped:

				this.p_ = this.r_ / this.m_ * 0.5  ;
				this.v_ = Math.sqrt( 4.0*this.m_*this.k_ - this.r_*this.r_ ) / this.m_ * 0.5 ;
				this.a_ = Math.sqrt( this.x0_*this.x0_ + this.v0_*this.v0_ / (this.p_*this.p_)) ;

				if (this.a_ == 0.0) {
					this.phi_ =	0.0 ;
				} else
				{
					this.phi_ =	Math.acos( this.x0_ / this.a_ ) ;
				}
				
				break ;
			
			case k_critically_damped: 
				this.c_ = this.x0_ ;
				this.b_ = this.v0_ + this.x0_ * this.w_ ;
				
				break ;
			
			default:
				alert("defa")
			
				break ;
		}

	},

	// Reference MST204/MST209 Handbook
	// 
	// 	centre()	-	returns the centre position of the spring
	// 
	// 	The centre is the position when the spring is in equilibrium.
	// 
	// public void centre( float & x ) // const
	// {
	// 	x = xe_ 
	// }
	
	centre: function()
	{ 
		return this.xe_ ;
	},
	

	// 
	// 	equilibrium()	-	check if the oscillator is in static equilibrium
	// 
	// 	The oscillator is static if it is at the equilibrium position with zero velocity.
	// 
	// 	Parameters:
	// 
	// 		t		-	the time to evaluate the system at
	// 		epsilon	-	how near to true equilbrium to calculate to
	// 
	// 	Returns:
	// 
	// 		the position of the system at the time t
	// 

	equilibrium: function( t,  epsilon ) 
	{ 
		return Math.abs( this.evaluate( t ) - this.xe_ ) <= epsilon && Math.abs( this.velocity(t ) ) <= epsilon  ;
	},
	
	

	
	evaluate: function( t) 
	{

		var x = 0.0
		var damping_type_ = this.damping_type_ 
		switch(damping_type_)
		{
			case k_strongly_damped:
				x = this.b_*Math.exp( this.l1_*t ) + this.c_*Math.exp( this.l2_*t ) 
				break ;
			case k_weakly_damped:
				x = this.a_*Math.exp( -this.p_*t ) * Math.cos( this.v_*t + this.phi_ ) 
				break ;
			
			case k_critically_damped:
				x = (this.b_*t +this.c_)*Math.exp( -this.w_*t ) 
				break ;
				
			default:
				break ;
		}
			
		return this.xe_ + x
	},
	
	velocity: function( t) 
	{

		var v = 0.0
		var damping_type_ = this.damping_type_ ;

		switch(damping_type_ )
		{
			case k_strongly_damped:
				v = this.b_*this.l1_*Math.exp( this.l1_*t ) + this.c_*this.l2_*Math.exp( this.l2_*t ) ;
				break ;
				 
			case k_weakly_damped: 
				v = -this.a_*this.p_*Math.exp( -this.p_*t ) * Math.cos( this.v_*t + this.phi_ ) - this.v_ * this.a_*Math.exp( -this.p_*t ) * Math.sin( this.v_*t + this.phi_ ) ;
				break ;
				 
			case k_critically_damped:
				v = ( this.b_ - this.w_*(this.b_*t +this.c_ ) )*Math.exp( -this.w_*t ) ;

			default:
				break ;
		}	

		return v 
	},
	
	toString: function()
	{
		return "KDampedOscillator" ;
	}
}
